contracts/Staking.sol

```diff
@@ -283,248 +338,241 @@ contract Staking is IStaking, System, IParamSubscriber, IApplication {
+      pendingRedelegateTime[delegator][valSrc][valDst] = block.timestamp.add(LOCK_TIME);
+      pendingRedelegateTime[delegator][valDst][valSrc] = block.timestamp.add(LOCK_TIME);
+
+      emit redelegateSuccess(delegator, valSrc, valDst, amount);
+    } else if (status == 0) {
+      pendingRedelegateTime[delegator][valSrc][valDst] = 0;
+      pendingRedelegateTime[delegator][valDst][valSrc] = 0;
+      emit redelegateFailed(delegator, valSrc, valDst, amount, errCode);
     } else {
-      require(false, "empty fail ack package");
+      revert("wrong status");
     }
+  }
 
-    bool success = false;
-    uint256 idx = 0;
+  function _handleRedelegateFailAckPackage(RLPDecode.Iterator memory paramIter) internal {
+    bool success;
+    uint256 idx;
     address delegator;
     address valSrc;
     address valDst;
     uint256 bcAmount;
-    while (iter.hasNext()) {
+    while (paramIter.hasNext()) {
       if (idx == 0) {
-        delegator = address(uint160(iter.next().toAddress()));
+        delegator = address(uint160(paramIter.next().toAddress()));
       } else if (idx == 1) {
-        valSrc = address(uint160(iter.next().toAddress()));
+        valSrc = address(uint160(paramIter.next().toAddress()));
       } else if (idx == 2) {
-        valDst = address(uint160(iter.next().toAddress()));
+        valDst = address(uint160(paramIter.next().toAddress()));
       } else if (idx == 3) {
-        bcAmount = uint256(iter.next().toUint());
+        bcAmount = uint256(paramIter.next().toUint());
         success = true;
       } else {
         break;
       }
-      idx++;
+      ++idx;
     }
-    require(success, "rlp decode package failed");
+    require(success, "rlp decode failed");
 
-    uint256 amount = bcAmount.mul(TEN_DECIMALS);
-    delegatedOfValidator[delegator][valSrc] = delegatedOfValidator[delegator][valSrc].add(amount);
-    delegatedOfValidator[delegator][valDst] = delegatedOfValidator[delegator][valDst].sub(amount);
+    redelegateInFly[delegator] -= 1;
     pendingRedelegateTime[delegator][valSrc][valDst] = 0;
     pendingRedelegateTime[delegator][valDst][valSrc] = 0;
 
     emit crashResponse(EVENT_REDELEGATE);
   }
 
   function _handleDistributeRewardSynPackage(RLPDecode.Iterator memory iter) internal returns(uint32, bytes memory) {
-    bool success = false;
-    uint256 idx = 0;
-    uint256 amount;
+    bool success;
+    uint256 idx;
     address recipient;
+    uint256 amount;
     while (iter.hasNext()) {
       if (idx == 0) {
-        amount = uint256(iter.next().toUint());
-      } else if (idx == 1) {
         recipient = address(uint160(iter.next().toAddress()));
+      } else if (idx == 1) {
+        amount = uint256(iter.next().toUint());
         success = true;
       } else {
         break;
       }
-      idx++;
+      ++idx;
     }
     if (!success) {
       return _encodeRefundPackage(EVENT_DISTRIBUTE_REWARD, amount, recipient, ERROR_FAIL_DECODE);
```

Comment:
if `!success`, it is possible that we did not get `amount` and `recipient`, should do revert here.

Reply:
Fixed